packages/svelte/src/compiler/utils/ast.js

```diff
@@ -54,41 +54,50 @@ export function is_event_attribute(attribute) {
 }
 
 /**
- * Extracts all identifiers from a pattern.
+ * Extracts all identifiers from a pattern. By default only those which can be newly declared as part of an assignment expression are included.
  * @param {ESTree.Pattern} param
  * @param {ESTree.Identifier[]} [nodes]
+ * @param {boolean} [include_member_expressions] If `true`, will also include member expressions which can be important in a case like `[a, b] = [c, d]`
  * @returns {ESTree.Identifier[]}
  */
-export function extract_identifiers(param, nodes = []) {
+export function extract_identifiers(param, nodes = [], include_member_expressions = false) {
 	switch (param.type) {
 		case 'Identifier':
 			nodes.push(param);
 			break;
 
+		case 'MemberExpression':
+			if (include_member_expressions) {
+				// Only the `a` from `[a.b[c].d]` is of interest to us here
+				const id = object(param);
+				if (id) nodes.push(id);
+			}
+			break;
```

Comment:
I don't understand what any of this is doing. If we revert all the changes in this file and keep the ones in `2-analyze/index.js`, the test still passes. Do we need it? It sets off my spidey senses

Reply:
Try running the whole test suite, you'll see one legacy test failing then

Reply:
I figured out what was triggering my spidey senses — this causes mutations and reassignments to be treated the same way, resulting in less efficient code in the `[array[1], array[0]] = [array[0], array[1]]` case. It's an edge case to be sure but I thought it worthwhile to get to the bottom of it anyway — opened #12429 with a very slightly different approach